In [3]:
#import relevant libraries and framework for analysis
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import matplotlib as mpl
import matplotlib.dates as mdates
import matplotlib.pylab as plb
import sweetviz as sv
import dtale
import plotly.express as px
import plotly.graph_objects as go
import plotly.figure_factory as ff
import plotly.offline as pyo
import plotly.io as pio
import folium
import folium.plugins as plugins
import folium.features as ft
import folium.vector_layers as vl
import folium.plugins as plugins
import seaborn as sns

C:\Users\DONKAMS\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


In [4]:
#load the dataset
dir =  'C:/Users/DONKAMS/Downloads/Stage-E-Exercise/amazon_books_2009-2019.csv'
df = pd.read_csv(dir, encoding='latin1')
df.head()

Unnamed: 0  price  rank                                          title  \
0           0  15.73   1.0               The Lost Symbol (Robert Langdon)   
1           1  10.14   2.0    The Shack: Where Tragedy Confronts Eternity   
2           2  14.76   3.0  Liberty and Tyranny: A Conservative Manifesto   
3           3  20.49   4.0      Breaking Dawn (The Twilight Saga, Book 4)   
4           4   6.30   5.0                  Going Rogue: An American Life   

   no_of_reviews  ratings            author cover_type  year  
0        29033.0      4.3         Dan Brown  Hardcover  2009  
1        41131.0      4.6  William P. Young  Paperback  2009  
2         5340.0      4.8     Mark R. Levin  Hardcover  2009  
3        25436.0      4.7   Stephenie Meyer  Hardcover  2009  
4         1598.0      4.6       Sarah Palin  Hardcover  2009

In [5]:
#rename index as S/N
df.index.name = 'S/N'
df.head()

Unnamed: 0  price  rank                                          title  \
S/N                                                                           
0             0  15.73   1.0               The Lost Symbol (Robert Langdon)   
1             1  10.14   2.0    The Shack: Where Tragedy Confronts Eternity   
2             2  14.76   3.0  Liberty and Tyranny: A Conservative Manifesto   
3             3  20.49   4.0      Breaking Dawn (The Twilight Saga, Book 4)   
4             4   6.30   5.0                  Going Rogue: An American Life   

     no_of_reviews  ratings            author cover_type  year  
S/N                                                             
0          29033.0      4.3         Dan Brown  Hardcover  2009  
1          41131.0      4.6  William P. Young  Paperback  2009  
2           5340.0      4.8     Mark R. Levin  Hardcover  2009  
3          25436.0      4.7   Stephenie Meyer  Hardcover  2009  
4           1598.0      4.6       Sarah Palin  Hardcover  2009

In [6]:
#remove the first column
df.drop('Unnamed: 0', axis=1, inplace=True)
df.head()

price  rank                                          title  \
S/N                                                               
0    15.73   1.0               The Lost Symbol (Robert Langdon)   
1    10.14   2.0    The Shack: Where Tragedy Confronts Eternity   
2    14.76   3.0  Liberty and Tyranny: A Conservative Manifesto   
3    20.49   4.0      Breaking Dawn (The Twilight Saga, Book 4)   
4     6.30   5.0                  Going Rogue: An American Life   

     no_of_reviews  ratings            author cover_type  year  
S/N                                                             
0          29033.0      4.3         Dan Brown  Hardcover  2009  
1          41131.0      4.6  William P. Young  Paperback  2009  
2           5340.0      4.8     Mark R. Levin  Hardcover  2009  
3          25436.0      4.7   Stephenie Meyer  Hardcover  2009  
4           1598.0      4.6       Sarah Palin  Hardcover  2009

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 898 entries, 0 to 897
Data columns (total 8 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   price          895 non-null    float64
 1   rank           895 non-null    float64
 2   title          892 non-null    object 
 3   no_of_reviews  895 non-null    float64
 4   ratings        895 non-null    float64
 5   author         893 non-null    object 
 6   cover_type     894 non-null    object 
 7   year           898 non-null    int64  
dtypes: float64(4), int64(1), object(3)
memory usage: 56.3+ KB


In [8]:
df.columns

Index(['price', 'rank', 'title', 'no_of_reviews', 'ratings', 'author',
       'cover_type', 'year'],
      dtype='object')

In [9]:
df.shape

(898, 8)

In [10]:
df.describe()

price        rank  no_of_reviews     ratings         year
count  895.000000  895.000000     895.000000  895.000000   898.000000
mean    13.548235   52.468156   31819.684425    4.610503  2014.121381
std      8.747561   31.286166   48000.086692    0.206606     3.210519
min      1.170000    1.000000       1.560000    3.500000  2009.000000
25%      8.890000   21.000000    6862.500000    4.500000  2011.000000
50%     12.150000   59.000000   17273.000000    4.600000  2014.000000
75%     15.735000   80.000000   35973.000000    4.800000  2017.000000
max     93.490000  100.000000  612377.000000    4.900000  2019.000000

#### To analyze the type of books frequently bought on Amazon between 2009 and 2019, you'll need to have data that includes at least the book type and the number of purchases for each book. Assuming you have a DataFrame df with columns like book_type, year, and number_of_purchases, here's a step-by-step plan:

- Filter the DataFrame for the years 2009 to 2019.
- Group the data by book_type and year, and sum the number_of_purchases.
- Sort the data by number_of_purchases in descending order to see the most frequently bought book types.

#### For this dataset, I can analyze the type of books  frequently bought on Amazon between 2009 and 2019 by using the cover_type as a proxy for book_type. However, I  don't have a direct measure of the number of purchases for each book. I might use no_of_reviews as a proxy, assuming that more reviews correlate with more purchases. Here's a step-by-step plan:

- Filter the DataFrame for the years 2009 to 2019.
- Group the data by cover_type and year, and sum the no_of_reviews.
- Sort the data by no_of_reviews in descending order to see the most frequently bought book types.
- Here's how you can do it in Python using pandas:

Next steps:

- Check the top_books DataFrame to see the results
- Visualize the results using a bar chart or other suitable plot
- Perform further analysis as needed, such as comparing the number of books displayed for sale with the number of purchases


In [11]:
#filter the dataset to include only the years 2009 to 2019
df_filtered = df[(df['year'] >= 2009) & (df['year'] <= 2019)]
df_filtered.head()

price  rank                                          title  \
S/N                                                               
0    15.73   1.0               The Lost Symbol (Robert Langdon)   
1    10.14   2.0    The Shack: Where Tragedy Confronts Eternity   
2    14.76   3.0  Liberty and Tyranny: A Conservative Manifesto   
3    20.49   4.0      Breaking Dawn (The Twilight Saga, Book 4)   
4     6.30   5.0                  Going Rogue: An American Life   

     no_of_reviews  ratings            author cover_type  year  
S/N                                                             
0          29033.0      4.3         Dan Brown  Hardcover  2009  
1          41131.0      4.6  William P. Young  Paperback  2009  
2           5340.0      4.8     Mark R. Levin  Hardcover  2009  
3          25436.0      4.7   Stephenie Meyer  Hardcover  2009  
4           1598.0      4.6       Sarah Palin  Hardcover  2009

 #### A new DataFrame df_filtered that only includes the rows where the year is between 2009 and 2019, inclusive.

In [15]:
df_grouped = df_filtered.groupby(['cover_type', 'year'])['no_of_reviews'].sum().reset_index()
df_grouped.sample(4)

cover_type  year  no_of_reviews
34              Paperback  2014       966173.0
31              Paperback  2011       592155.0
26  Mass Market Paperback  2017       110385.0
27  Mass Market Paperback  2018        94095.0

In [17]:
# Get unique years
unique_years = df['year'].unique()

# Create a dictionary to store each year's DataFrame
dfs_by_year = {}

for year in unique_years:
    # Filter df for the specific year
    dfs_by_year[year] = df[df['year'] == year]
#list of years
list(dfs_by_year.keys())

[2009, 2010, 2011, 2012, 2013, 2014, 2015, 2016, 2017, 2018, 2019]

#### So we now have a list of filtered dataframe for each year 

In [18]:
#lets see for 2009 now
dfs_by_year[2009].head()

price  rank                                          title  \
S/N                                                               
0    15.73   1.0               The Lost Symbol (Robert Langdon)   
1    10.14   2.0    The Shack: Where Tragedy Confronts Eternity   
2    14.76   3.0  Liberty and Tyranny: A Conservative Manifesto   
3    20.49   4.0      Breaking Dawn (The Twilight Saga, Book 4)   
4     6.30   5.0                  Going Rogue: An American Life   

     no_of_reviews  ratings            author cover_type  year  
S/N                                                             
0          29033.0      4.3         Dan Brown  Hardcover  2009  
1          41131.0      4.6  William P. Young  Paperback  2009  
2           5340.0      4.8     Mark R. Levin  Hardcover  2009  
3          25436.0      4.7   Stephenie Meyer  Hardcover  2009  
4           1598.0      4.6       Sarah Palin  Hardcover  2009

In [19]:
#check the shape of the 2009 dataset
dfs_by_year[2009].shape 

(79, 8)

<h1 align = 'center'> Analysis of Top book reviews by year </h1>

In [21]:
df_sorted = dfs_by_year[2009].sort_values('no_of_reviews', ascending=False)
df_sorted.head()

price  rank                                              title  \
S/N                                                                   
54   10.30  75.0  How to Win Friends & Influence People (Dale Ca...   
38    7.83  59.0                                     The Book Thief   
60   22.60  81.0                                   The Hunger Games   
1    10.14   2.0        The Shack: Where Tragedy Confronts Eternity   
18   10.32  19.0  The Guernsey Literary and Potato Peel Pie Society   

     no_of_reviews  ratings            author cover_type  year  
S/N                                                             
54        111939.0      4.7     Dale Carnegie  Paperback  2009  
38         67291.0      4.6      Markus Zusak  Paperback  2009  
60         42770.0      4.7   Suzanne Collins  Hardcover  2009  
1          41131.0      4.6  William P. Young  Paperback  2009  
18         37114.0      4.6  Mary Ann Shaffer  Paperback  2009

### For the year 2009, The book How to Win Friends & Influence People has the highest number of reviews with an high rating of 4.7, The author is Dale Carnegie, which has a paperback cover type, Most of the books having the highest reviews also have an high rating between 4.6 and 4.7, hence there is a strong corellation between the number of reviews and ratings, Thou there ranking seems bit away from the top ranked

#### Also the top five authors by the number of reviews are
- Dale Carnegie
- Markus Zusak
- Suzanne Collins
- William P. Young
- Mary Ann Shaffer

#### The top five top books are;
- How to Win Friends & Influence People (Dale Ca
- The Book Thief
- The Hunger Games	
- The Shack: Where Tragedy Confronts Eternity	
- The Guernsey Literary and Potato Peel Pie Society	

In [22]:
df_sorted = dfs_by_year[2010].sort_values('no_of_reviews', ascending=False)
df_sorted.head()

price  rank                                              title  \
S/N                                                                   
86   11.44   8.0  Mockingjay (the Final Book of the Hunger Games...   
92   15.68  14.0  Unbroken: A World War II Story of Survival, Re...   
106  11.64  28.0                   Catching Fire (The Hunger Games)   
114  20.64  56.0               Fall of Giants (The Century Trilogy)   
131   8.89  73.0  Jesus Calling, Padded Hardcover, with Scriptur...   

     no_of_reviews  ratings             author cover_type  year  
S/N                                                              
86         79880.0      4.6    Suzanne Collins  Hardcover  2010  
92         75437.0      4.7  Laura Hillenbrand  Hardcover  2010  
106        64916.0      4.7    Suzanne Collins  Hardcover  2010  
114        50707.0      4.4        Ken Follett  Hardcover  2010  
131        43017.0      4.9        Sarah Young  Hardcover  2010

#### The book Mockingjay (the Final Book of the Hunger Games.. has the highest review in 2010 with 79.8k reviews and a rating of 4.4 to 4.9 surrounding the top 5 books, Unlike the Topp 5 books in 2009, All top 5 in this year have an hardcover as there cover type. Also the Top 5 author includes: 
- Suzanne Collins
- Laura Hillenbrand
- Ken Follett
- Sarah Young

#### It is worthy of note to state that Suzanne Collins has two of her books in the top review for this year (2010), she also made it to the top 5 in 2009
#### The following are the books that made top 5;
- Mockingjay (the Final Book of the Hunger Games...	
- Unbroken: A World War II Story of Survival, Re...	
- Catching Fire (The Hunger Games)	
- Fall of Giants (The Century Trilogy)	
- Jesus Calling, Padded Hardcover, with Scriptur...	


In [24]:
df_sorted = dfs_by_year[2011].sort_values('no_of_reviews', ascending=False)
df_sorted.head()

price  rank                                              title  \
S/N                                                                   
215  45.44  78.0         Harry Potter Paperback Box Set (Books 1-7)   
162   7.19   5.0                          The Hunger Games (Book 1)   
166  11.44   9.0  Mockingjay (the Final Book of the Hunger Games...   
163  15.68   6.0  Unbroken: A World War II Story of Survival, Re...   
204   7.83  67.0                                     The Book Thief   

     no_of_reviews  ratings             author cover_type  year  
S/N                                                              
215       112398.0      4.9      J. K. Rowling  Paperback  2011  
162        80816.0      4.7    Suzanne Collins  Paperback  2011  
166        79880.0      4.6    Suzanne Collins  Hardcover  2011  
163        75437.0      4.7  Laura Hillenbrand  Hardcover  2011  
204        67292.0      4.6       Markus Zusak  Paperback  2011

### Suzanne collins has made it having two of her books in the top review for this year also Her book Mockingjay (the Final Book of the Hunger Games...	has succedded in having a consecutive top streak. One notable personality is also Laura Hillenbrand she has also made the list after been among the top author in the previous year (2010) . Markus Zusak is also a personality to watch out for. he was featured in 2009 as one of the top author with the most reviewed books, Most books in this year have there cover type to be Paperback

In [25]:
df_sorted = dfs_by_year[2012].sort_values('no_of_reviews', ascending=False)
df_sorted.head()

price  rank                                              title  \
S/N                                                                   
249  16.98  12.0                                          Gone Girl   
238  11.13   1.0  Fifty Shades Of Grey: Book One of the Fifty Sh...   
280   7.74  63.0  The Four Agreements: A Practical Guide to Pers...   
241   7.19   4.0                          The Hunger Games (Book 1)   
245  11.44   8.0  Mockingjay (the Final Book of the Hunger Games...   

     no_of_reviews  ratings           author cover_type  year  
S/N                                                            
249       161568.0      4.1    Gillian Flynn  Hardcover  2012  
238       111859.0      4.2        E L James  Paperback  2012  
280       107993.0      4.7  Don Miguel Ruiz  Paperback  2012  
241        80816.0      4.7  Suzanne Collins  Paperback  2012  
245        79880.0      4.6  Suzanne Collins  Hardcover  2012

#### It interesting to note that the top ranked book 'Fifty Shades of Grey' is the second most reviewedin this year (2012)

#### It's 2012 and Suzanne Collins has proven her worth again. Her two books (The Hunger Games (Book 1) and Mockingjay (the Final Book of the Hunger Games... ) have been featured among the top 5 reviews, Interestinly,the book Mockingjay (the Final Book of the Hunger Games...	has sucessfully featured as a top 5 item in the past 3 years with The Hunger Games (Book 1) having present in top 5 of 2011 and 2009. Most top books have there cover type to be Paperback. Gillian Flynn book Gone Girl has the highest recorded number of reviews since 2009 161.6K

In [26]:
df_sorted = dfs_by_year[2013].sort_values('no_of_reviews', ascending=False)
df_sorted.head()

price  rank                                              title  \
S/N                                                                   
331  10.80  14.0                             The Fault in Our Stars   
347  18.50  30.0  The Goldfinch: A Novel (Pulitzer Prize for Fic...   
373  10.30  76.0  How to Win Friends & Influence People (Dale Ca...   
333   9.40  16.0                                             Wonder   
323  15.69   6.0                  Inferno: A Novel (Robert Langdon)   

     no_of_reviews  ratings         author cover_type  year  
S/N                                                          
331       160029.0      4.6     John Green  Hardcover  2013  
347       124205.0      4.1    Donna Tartt  Hardcover  2013  
373       111939.0      4.7  Dale Carnegie  Paperback  2013  
333        83334.0      4.7  R. J. Palacio  Hardcover  2013  
323        82446.0      4.2      Dan Brown  Hardcover  2013

#### Dale Carnegie was the top author in 2010 with this same book How to Win Friends & Influence People (Dale Ca...with	 the same ratings,most top reviewed book this year(2013) have cover type to be Hardcover.

In [27]:
df_sorted = dfs_by_year[2014].sort_values('no_of_reviews', ascending=False)
df_sorted.head()

price  rank                                              title  \
S/N                                                                   
414  12.90  17.0                                          Gone Girl   
401  11.69   4.0                             The Fault in Our Stars   
411  10.80  14.0                             The Fault in Our Stars   
413  20.59  16.0                              To Kill a Mockingbird   
418  18.50  21.0  The Goldfinch: A Novel (Pulitzer Prize for Fic...   

     no_of_reviews  ratings         author             cover_type  year  
S/N                                                                      
414       161570.0      4.1  Gillian Flynn              Paperback  2014  
401       160029.0      4.6     John Green              Paperback  2014  
411       160029.0      4.6     John Green              Hardcover  2014  
413       130637.0      4.7     Harper Lee  Mass Market Paperback  2014  
418       124205.0      4.1    Donna Tartt              Hardcover  2014

#### The year 2014 is an interesting one, The Gilian Flynn book has topped the table for the second time after 2012, this time the number of review only increased by only 2 reviews with the over type actually been changed from Hardcover and Paperback, this might be the reasone for low increase in the rating and review. The book 'The fault in our stars' has both the two cover type Paperback and Hardcover to feature in the top 5, kudos to John Green for the good work.

In [34]:
df_sorted = dfs_by_year[2015].sort_values('no_of_reviews', ascending=False)
df_sorted.head()

price  rank                                              title  \
S/N                                                                   
486  11.11   9.0                              The Girl on the Train   
485  15.15   8.0                        All the Light We Cannot See   
493  10.41  16.0                                        The Martian   
489  20.59  12.0                              To Kill a Mockingbird   
501  12.76  24.0  Grey: Fifty Shades of Grey as Told by Christia...   

     no_of_reviews  ratings         author             cover_type  year  
S/N                                                                      
486       492085.0      4.1  Paula Hawkins              Hardcover  2015  
485       215273.0      4.5  Anthony Doerr              Hardcover  2015  
493       170736.0      4.6      Andy Weir              Paperback  2015  
489       130637.0      4.7     Harper Lee  Mass Market Paperback  2015  
501       117817.0      4.5      E L James              Paperback  2015

#### 2015 is a year with a different beahviour, Most authors topping the list now are new and have there books ranked in the Top 30 for the year, Notable Harper Lee takes on the mantle as he is featured again in the table for the second time consecutively. 'The girl on the train', a book written by Paula Hawkins has achieved the highest review since 2009 with 492.1k reviews The is a very tremendous achievement for a book that has ot made it to the Top 5 before.

In [35]:
df_sorted = dfs_by_year[2016].sort_values('no_of_reviews', ascending=False)
df_sorted.head()

price  rank                                              title  \
S/N                                                                   
577  10.33  20.0                              The Girl on the Train   
592  15.15  55.0                        All the Light We Cannot See   
568  10.93  11.0                          A Man Called Ove: A Novel   
594  11.15  57.0  The Alchemist, 25th Anniversary: A Fable About...   
573  20.59  16.0                              To Kill a Mockingbird   

     no_of_reviews  ratings           author             cover_type  year  
S/N                                                                        
577       492085.0      4.1    Paula Hawkins              Paperback  2016  
592       215273.0      4.5    Anthony Doerr              Hardcover  2016  
568       186212.0      4.6  Fredrik Backman              Paperback  2016  
594       140829.0      4.6     Paulo Coelho              Paperback  2016  
573       130637.0      4.7       Harper Lee  Mass Market Paperback  2016

#### Continuing on the magic streak, the Paula Hawkins book "The girl on Titans" has made it to the top again with the same number of reviewss and rating as the previous year, However it is notable to say that the price was reduced from 11.11 to 10.33 which causes ut ranking to drop from 9 to 20. Harper Lee , a popular author with a popular rating also made it to the list with the same book 'To kill a Mocking bird'

In [30]:
df_sorted = dfs_by_year[2017].sort_values('no_of_reviews', ascending=False)
df_sorted.head()

price  rank                                              title  \
S/N                                                                   
657  10.93  20.0                          A Man Called Ove: A Novel   
670  13.07  53.0                                    Big Little Lies   
648   9.53  11.0                                The Handmaid's Tale   
684  11.15  67.0  The Alchemist, 25th Anniversary: A Fable About...   
641  14.38   4.0  The Subtle Art of Not Giving a F*ck: A Counter...   

     no_of_reviews  ratings           author cover_type  year  
S/N                                                            
657       186212.0      4.6  Fredrik Backman  Paperback  2017  
670       157430.0      4.4   Liane Moriarty  Paperback  2017  
648       154708.0      4.3  Margaret Atwood  Paperback  2017  
684       140829.0      4.6     Paulo Coelho  Paperback  2017  
641       138110.0      4.6      Mark Manson  Hardcover  2017

#### 2017, a year where all top review books have there cover type to be Paperback and also a close no of reviews between each book . 

In [31]:
df_sorted = dfs_by_year[2018].sort_values('no_of_reviews', ascending=False)
df_sorted.head()

price  rank                           title  no_of_reviews  ratings  \
S/N                                                                        
767  14.67  70.0         Where the Crawdads Sing       612377.0      4.7   
757  13.46  60.0  Beneath a Scarlet Sky: A Novel       221991.0      4.6   
734  14.66  17.0              Educated: A Memoir       208462.0      4.6   
718  17.18   1.0                        Becoming       195572.0      4.8   
754  14.73  57.0   Before We Were Yours: A Novel       141704.0      4.6   

                 author cover_type  year  
S/N                                       
767         Delia Owens  Hardcover  2018  
757       Mark Sullivan  Paperback  2018  
734       Tara Westover  Hardcover  2018  
718      Michelle Obama  Hardcover  2018  
754  Dungeons & Dragons  Hardcover  2018

#### This is interesting, The record for the highest number of reviews renewed with 612.4k for Delia Owens book 'Where the Crawdads sing'. All top five books have a very high rating between 4.6- 4.8, The is the highest recorded. The best book for this year Becoming of Michelle Obama is also one of the top five most reviewed. THis is also the first time since 2009 that the most ranked book will be featured in the top 5. The books are also in Hardcover cover type.

In [32]:
df_sorted = dfs_by_year[2019].sort_values('no_of_reviews', ascending=False)
df_sorted.head()

price  rank                                              title  \
S/N                                                                   
798  12.39   1.0                            Where the Crawdads Sing   
860  13.99  63.0                            Where the Crawdads Sing   
836  12.99  39.0                                 The Silent Patient   
799  16.50   2.0                                           Becoming   
862  13.29  65.0  The Alchemist, 25th Anniversary: A Fable About...   

     no_of_reviews  ratings            author cover_type  year  
S/N                                                             
798       344811.0      4.8       Delia Owens  Hardcover  2019  
860       344811.0      4.8       Delia Owens  Paperback  2019  
836       135163.0      4.5  Alex Michaelides  Hardcover  2019  
799       123276.0      4.8    Michelle Obama  Hardcover  2019  
862        94750.0      4.7      Paulo Coelho  Paperback  2019